In [2]:
%load_ext kedro.ipython

[07/23/24 17:32:02] INFO     Registered line magic '%reload_kedro'                                   ]8;id=675409;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=12619;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#58\58]8;;\

                    INFO     Registered line magic '%load_node'                                      ]8;id=852292;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=944492;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#60\60]8;;\

                    INFO     Resolved project path as:                                              ]8;id=722566;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=236551;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#171\171]8;;\
                             c:\Users\PC\Desktop\PRO-G\ECOLE\DSTI_goodreads\goodreads-predictor.                   
                             To set a different path, run '%reload_kedro <project_root>'                           

[07/23/24 17:32:07] INFO     Kedro project goodreads-predictor                                      ]8;id=598771;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=313296;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#141\141]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=797259;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=761064;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#142\142]8;;\
                             'pipelines'                                                                           

[07/23/24 17:32:09] INFO     Registered line magic 'run_viz'                                        ]8;id=343473;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=572802;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#148\148]8;;\

In [3]:
%reload_kedro

                    INFO     Resolved project path as:                                              ]8;id=369761;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=807755;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#171\171]8;;\
                             C:\Users\PC\Desktop\PRO-G\ECOLE\DSTI_goodreads\goodreads-predictor.                   
                             To set a different path, run '%reload_kedro <project_root>'                           

[07/23/24 17:32:10] INFO     Kedro project goodreads-predictor                                      ]8;id=709159;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=649719;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#141\141]8;;\

                    INFO     Defined global variable 'context', 'session', 'catalog' and            ]8;id=843821;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=663593;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#142\142]8;;\
                             'pipelines'                                                                           

                    INFO     Registered line magic 'run_viz'                                        ]8;id=377312;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py\__init__.py]8;;\:]8;id=555705;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\ipython\__init__.py#148\148]8;;\

In [4]:
import pandas as pd
import re
import numpy as np
import matplotlib.pyplot as plt
import plotly.graph_objects as go # For the waterfall plot -- cannot easily do this in matplotlib
from typing import List, Dict, Tuple, Any



In [5]:
context.catalog.list()


[
    'books_raw',
    'books_loaded',
    'publisher_consolidation',
    'price_by_isbn_input',
    'price_by_isbn',
    'title_embeddings_original',
    'authors_embeddings',
    'description_embeddings',
    'open_library_book_api_info_input',
    'open_library_book_api_info',
    'filtered_books',
    'exclusions_summary',
    'scope_waterfall_plot',
    'parameters',
    'params:color_gr_brown',
    'params:color_gr_purple',
    'params:color_gr_green',
    'params:color_gr_tan_background',
    'params:title_column',
    'params:author_column',
    'params:description_column',
    'params:original_data_key_column'
]

In [26]:
books_input = catalog.load('books_loaded')
price_by_isbn = catalog.load('price_by_isbn')
title_embeddings = catalog.load('title_embeddings_original')
book_api_data = catalog.load('open_library_book_api_info')
authors_embeddings = catalog.load('authors_embeddings')

# Get the color values
color_gr_brown = catalog.load("params:color_gr_brown")
color_gr_purple = catalog.load("params:color_gr_purple")
color_gr_green = catalog.load("params:color_gr_green")
color_gr_tan_background = catalog.load("params:color_gr_tan_background")

# Add in the price data
books_input = books_input.merge(price_by_isbn, on='isbn13', how='left')

[07/23/24 17:45:15] INFO     Loading data from books_loaded (ParquetDataset)...                 ]8;id=498577;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=702338;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from price_by_isbn (ParquetDataset)...                ]8;id=483061;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=694926;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from title_embeddings_original (ParquetDataset)...    ]8;id=640341;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=160871;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from open_library_book_api_info (PickleDataset)...    ]8;id=265067;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=359920;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_brown (MemoryDataset)...         ]8;id=912210;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=812425;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_purple (MemoryDataset)...        ]8;id=331922;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=231667;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_green (MemoryDataset)...         ]8;id=542431;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=279447;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\

                    INFO     Loading data from params:color_gr_tan_background                   ]8;id=154769;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py\data_catalog.py]8;;\:]8;id=955471;file://c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\kedro\io\data_catalog.py#508\508]8;;\
                             (MemoryDataset)...                                                                    

╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:14                                                                                   │
│                                                                                                  │
│   11                                                                                             │
│   12 # Add in the price data                                                                     │
│   13 books_input = books_input.merge(price_by_isbn, on='isbn13', how='left')                     │
│ ❱ 14 books_input = books_input.merge(authors_embeddings, on='authors')                           │
│   15                                                                                             │
│                                                                                                  │
│ c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\pandas\core\frame.py:10832 in merge         │
│                                                                                                  │
│ c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\pandas\core\reshape\merge.py:170 in merge   │
│                                                                                                  │
│ c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\pandas\core\reshape\merge.py:794 in         │
│ __init__                                                                                         │
│                                                                                                  │
│ c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\pandas\core\reshape\merge.py:1297 in        │
│ _get_merge_keys                                                                                  │
│                                                                                                  │
│ c:\Users\PC\.conda\envs\goodreadv2\lib\site-packages\pandas\core\generic.py:1911 in              │
│ _get_label_or_level_values                                                                       │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
KeyError: 'authors'

# Purpose of this notebook 

## This NB is an exploratory data analysis part of the project.
We'll suggest some hypothesis and try to verify them through analysis 

In [7]:
data = books_input

In [29]:
df = data.merge(title_embeddings)


In [30]:
df = data.merge(authors_embeddings)

In [31]:
df.head(5)

bookID                                              title  \
0       1  Harry Potter and the Half-Blood Prince (Harry ...   
1       2  Harry Potter and the Order of the Phoenix (Har...   
2       4  Harry Potter and the Chamber of Secrets (Harry...   
3       5  Harry Potter and the Prisoner of Azkaban (Harr...   
4       8  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   

                      authors  average_rating        isbn         isbn13  \
0  J.K. Rowling/Mary GrandPré            4.57  0439785960  9780439785969   
1  J.K. Rowling/Mary GrandPré            4.49  0439358078  9780439358071   
2                J.K. Rowling            4.42  0439554896  9780439554893   
3  J.K. Rowling/Mary GrandPré            4.56  043965548X  9780439655484   
4  J.K. Rowling/Mary GrandPré            4.78  0439682584  9780439682589   

  language_code    num_pages  ratings_count  text_reviews_count  \
0           eng          652        2095690               27591   
1           eng          870        2153167               29221   
2           eng          352           6333                 244   
3           eng          435        2339585               36325   
4           eng         2690          41428                 164   

  publication_date        publisher  Price PriceStrategy  \
0        9/16/2006  Scholastic Inc.  12.99    LIST_PRICE   
1         9/1/2004  Scholastic Inc.  99.99    LIST_PRICE   
2        11/1/2003       Scholastic  24.95    LIST_PRICE   
3         5/1/2004  Scholastic Inc.   7.99    LIST_PRICE   
4        9/13/2004       Scholastic  40.95    LIST_PRICE   

                                   authors_embedding  
0  [-0.30315670371055603, 0.11502666771411896, -0...  
1  [-0.30315670371055603, 0.11502666771411896, -0...  
2  [-0.22978533804416656, 0.1457509994506836, -0....  
3  [-0.30315670371055603, 0.11502666771411896, -0...  
4  [-0.30315670371055603, 0.11502666771411896, -0...

In [72]:
df['authors'].nunique()

6639

# First hypothesis 

By design, an author might be related to the rating of a book. For example, Charles Dickens wrote classics that might be just all rated the same. So the author's name's by default a significant element to the average rating of a book. Let's try and proove this assumption 

In [33]:
first_hypodf=df[['authors', 'average_rating', 'authors_embedding']]

In [34]:
first_hypodf.describe()

average_rating
count    11123.000000
mean         3.934075
std          0.350485
min          0.000000
25%          3.770000
50%          3.960000
75%          4.140000
max          5.000000

## First thing we notice 
Those books aren't very bad! As the mean is 4 out of 5, we can assume that we might not have much many lame books... That's a first bias we definitely need to handle

## Hypothesis 1.a / Embeddings catch correct relations

By computing the cosine similarity we can prove or forget the embedding technique

In [39]:
from sklearn.metrics.pairwise import cosine_similarity

In [44]:
"""
Trying to see if the embedding catches similarities between different instances of a same author
"""

one = first_hypodf['authors_embedding'].iloc[0]
two = first_hypodf['authors_embedding'].iloc[2]

In [52]:
one = one.reshape(1,-1)
two = two.reshape(1,-1)

In [53]:
cosine_similarity(one,two)

array([[0.75045805]])

A cosine similarity of 0.75 is very good for two instances fairly different. We can use this technique to merge names together. For example we can set a threshold of 70% and basically say that if the cosine similarity of two embeddings is greater or equal to 70 the author is the same or has some differences (a collaborator like the Mary GrandPré which is the illustrator of the latests HP books.)

In [56]:
# Basically converting all authors embedding to match a cosine computation

first_hypodf.loc[:, 'authors_embedding'] = first_hypodf['authors_embedding'].apply(lambda x: x.reshape(1, -1))

In [66]:
def standardize_author_names(df, threshold=0.72):
    embeddings = first_hypodf['authors_embedding'].tolist()
    authors = first_hypodf['authors'].tolist()
    
    
    embeddings = [embedding.flatten() for embedding in embeddings]
    
    for i in range(len(embeddings)):
        current_embedding = embeddings[i].reshape(1, -1)
        similarities = cosine_similarity(current_embedding, embeddings)
        
        for j in range(len(similarities[0])):
            if i != j and similarities[0][j] >= threshold:
                authors[i] = authors[j]
                break
    
    first_hypodf.loc[:, 'authors'] = authors

In [67]:
standardize_author_names(first_hypodf)

In [71]:
first_hypodf['authors'].nunique()

1567

We went from 6639 authors name to 1567 in a robust manner this is actually very good to analyse with ease 